Use user identify from $LIMA_HOME/_config/user

jandubois opened the pull request:
By default lima uses the public keys from `~/.ssh/*.pub`.

This change makes it try `$LIME_HOME/_config/user.pub` first and only fall back on the `~/.ssh/*.pub` keys when that one does not exist. It will create the key under `$LIMA_HOME/_config` if there aren't any public keys under `~/.ssh`.

Fixes #78

AkihiroSuda commented on pkg/sshutil/sshutil.go:
```diff
@@ -18,34 +19,74 @@ type PubKey struct {
+	content, err := os.ReadFile(f)
+	if err == nil {
+		entry.Content = strings.TrimSpace(string(content))
+	} else {
```
Return err if !error.Is(err, os.ErrNorExist)

jandubois replied:
Why? If `ReadFile` returns an error, `content` cannot be used, and we should return the error, regardless of what it is.

It will just be written to the log, and then ignored because `DefaultPubKeys` doesn't return an error. Maybe it should?

jandubois replied:
This is now handled by the caller; any error is reported back to the caller because we are only calling `readPublicKey` on files that are supposed to exist.

AkihiroSuda commented on pkg/sshutil/sshutil.go:
```diff
@@ -18,34 +19,74 @@ type PubKey struct {
-		return nil
+	} else {
+		// Check that private key exists and then try to read public key
+		if _, err := os.Stat(privateKeyPath); err == nil {
```
Return err if !error.Is(err, os.ErrNorExist)

jandubois replied:
`DefaultPubKeys` doesn't return any errors; it just ignores key files it cannot read. So not sure what to do here.

In the end the user would be told to run `ssh-keygen` when `DefaultPubKeys` returns an empty list, so not sure that aborting on the first error is really helpful.

AkihiroSuda commented on pkg/sshutil/sshutil.go:
```diff
@@ -83,5 +124,12 @@ func SSHArgs(instDir string) ([]string, error) {
 		"-o", "BatchMode=yes",
 	}
+	configDir, err := store.LimaConfigDir()
+	if err == nil {
```
Return err if !error.Is(err, os.ErrNorExist)

AkihiroSuda commented on pkg/sshutil/sshutil.go:
```diff
@@ -83,5 +124,12 @@ func SSHArgs(instDir string) ([]string, error) {
+	configDir, err := store.LimaConfigDir()
+	if err == nil {
+		privateKeyPath := filepath.Join(configDir, "user")
+		if _, err := os.Stat(privateKeyPath); err == nil {
```
Return err if !error.Is(err, os.ErrNorExist)

AkihiroSuda commented on pkg/sshutil/sshutil.go:
```diff
@@ -18,34 +19,74 @@ type PubKey struct {
+	if configDir == "" {
+		return nil
+	}
+	keygenCmd := exec.Command("ssh-keygen", "-q", "-N", "", "-f", privateKeyPath)
```
Can we specify algo explicitly?

AkihiroSuda replied:
Perhaps ed25519 is fine

AkihiroSuda commented on pkg/sshutil/sshutil.go:
```diff
@@ -18,34 +19,74 @@ type PubKey struct {
+		return nil
+	}
+	keygenCmd := exec.Command("ssh-keygen", "-q", "-N", "", "-f", privateKeyPath)
+	if out, err := keygenCmd.CombinedOutput(); err != nil {
```
logrus.Debugf(“executing %v”, keygenCmd.Args)

AkihiroSuda commented on pkg/store/store.go:
```diff
@@ -29,6 +29,20 @@ func LimaDir() (string, error) {
 }
 
+// LimaConfigDir returns the path of the config directory, $LIMA_HOME/_config.
+// It will create the directory if it doesn't yet exist.
```
I’m not sure this function should make the dir.

jandubois replied:
Yeah, it is inconsistent with what `LimaDir` does, but I was thinking about letting that one create the directory too. Would that simplify the calling code?

AkihiroSuda commented on pkg/store/store.go:
```diff
@@ -29,6 +29,20 @@ func LimaDir() (string, error) {
+	if err != nil {
+		return "", err
+	}
+	configDir := filepath.Join(limaDir, "_config")
```
“_config” should be a constant.

jandubois replied:
Yes, I was tempted to put it into `store/filenames`, but that package claims it only defines names for the instance directories. So where should `_config` be defined? New package?

AkihiroSuda replied:
pkg/store.ConfigDir

jandubois commented:
@AkihiroSuda I believe I've addresses all issues you pointed out. Can you please confirm?

But I'm wondering now if the `user` filename shouldn't be a constant too, but again don't know where to put it.

Should `store/filenames` be extended to cover both config and instance filenames?

jandubois commented:
Sorry, I refactored `DefaultPubKeys` once more, so that it can actually reflect an error back to the caller.

By tightening the error handling I've also reduced the logic nesting level a bit, as it was getting out of hand.

I've also gone ahead and added other file and directory names to the `filenames` package. Let me know, if this is not how you want it to be!

I believe I'm done with it now (unless you have more feedback).

jandubois commented:
@AkihiroSuda The latest CI failure is different, but I've seen a lot of errors like this before:

```
time="2021-06-28T21:49:58Z" level=fatal msg="failed to resolve reference \"ghcr.io/stargz-containers/nginx:1.19-alpine-org\": failed to do request: Head \"https://ghcr.io/v2/stargz-containers/nginx/manifests/1.19-alpine-org\": net/http: TLS handshake timeout"
```

Any idea why that is happening? You would think the connection between gha and ghcr should be pretty solid...

AkihiroSuda commented:
I guess macOS instances might be running on a different infra that isn't optimal for ghcr?

AkihiroSuda commented on pkg/sshutil/sshutil.go:
```diff
@@ -19,34 +20,71 @@ type PubKey struct {
+		if err := os.MkdirAll(configDir, 0700); err != nil {
+			return nil, errors.Wrapf(err, "could not create %q directory", configDir)
+		}
+		keygenCmd := exec.Command("ssh-keygen", "-t", "ed25519", "-q", "-N", "", "-f",
```
We should lock the directory. I can open a follow-up PR.

AkihiroSuda commented on pkg/sshutil/sshutil.go:
```diff
@@ -77,7 +115,17 @@ func SSHArgs(instDir string) ([]string, error) {
+		return nil, err
+	}
 	args := []string{
+		"-i", privateKeyPath,
```
Question: would it be possible to let ssh read both `~/.lima/_config` keys and `~/.ssh` keys?

Otherwise this will break existing instances.

jandubois replied:
> Question: would it be possible to let ssh read both `~/.lima/_config` keys and `~/.ssh` keys?

Yes, you can specify multiple `-i` options, and the identities will be tried in sequence. I think the server may be limiting the number of attempts to 3, so it could still fail if the user added multiple keys after creating the instance. But that is already a problem with the current implementation, so doesn't change anything. Just another reason to have a lima-specific identity, and always specify that one first.

jandubois replied:
> Question: would it be possible to let ssh read both `~/.lima/_config` keys and `~/.ssh` keys?

I've pushed the change in a separate commit for easier review; lmk if you want me to squash it!

AkihiroSuda reviewed (approved):
Thanks
